### ISSUE TITLE
Add subjects to Toots

### ISSUE THREAD

**Issue by nothingneko** [mastodon/mastodon#25125]
### Pitch

I've noticed that a lot of people use Content Warnings to denote what content they're posting about. This has always felt like it devalued Content Warnings. I propose adding a subject line like with email to toots, it would allow users to denote what their post was about without hiding the contents.

### Motivation

I feel the widespread usage of Content Warnings to denote the topic of a post devalues Content Warnings, and makes the feed harder to navigate on mobile. I think this would benefit those who add Content Warnings to their safe posts.

**Comment by trwnh** [comment]
i think this is possible and could be done, but it would be an awkward transitionary period where some servers/softwares are aware of the new property and some aren't (if you want to be able to set a "subject" and a "content warning" separately, that is)

basically, if you want to set both a subject and a cw, you have to decide which one will "take over" from the existing `summary` property, and which one gets its own property. i think the cleanest thing to do would be to define a `contentWarning` property.

alternatively, if you are fine with only being able to set one of them at a time: we could treat `summary` differently depending on whether the status is `sensitive` or not.

**Comment by nothingneko** [comment]
I think the last option there makes the most sense, when a content warning is needed it acts as the subject

**Comment by joyeusenoelle** [comment]
I agree with the last option as well, I think. In my opinion Mastodon should shift its terminology so that the existing "Content Warning" is renamed "Subject", which can be treated as a content warning if necessary.

**Comment by nothingneko** [comment]
If you were to do that, you'd also have to make hiding optional, so you can still hide sensitive topics

**Comment by joyeusenoelle** [comment]
I think one of us has misunderstood; I'm not suggesting removing the hiding feature. In fact, that remains a substantial benefit as far as I'm concerned; if someone's talking about a TV show I don't watch, I don't need a content warning for it, but I'm still not going to click the "show more" to open the full post, much like if I read the subject line of an email and decide I don't care about it, I won't click to open the email. My suggestion is literally "rename CW to Subject, change nothing else, and then people can use the subject for a CW if they want to". This retains as much of current cultural usage as possible, is a minimal change to the codebase, and doesn't devalue the idea of CWs by using something explicitly called a content warning as a mere subject line.

**Comment by nothingneko** [comment]
Ah, understood

**Comment by JupiterRowland** [comment]
> i think this is possible and could be done, but it would be an awkward transitionary period where some servers/softwares are aware of the new property and some aren't (if you want to be able to set a "subject" and a "content warning" separately, that is)
> 
> basically, if you want to set both a subject and a cw, you have to decide which one will "take over" from the existing `summary` property, and which one gets its own property. i think the cleanest thing to do would be to define a `contentWarning` property.
> 
> alternatively, if you are fine with only being able to set one of them at a time: we could treat `summary` differently depending on whether the status is `sensitive` or not.

Many other Fediverse projects, including some that predate Mastodon by years (Friendica, Hubzilla), have both a subject field and a summary field. This works fairly well when posts from these projects arrive on Mastodon: The subject is discarded, the summary is understood as a content warning because it's technically the same anyway. No hiccups because of this in seven years AFAIK.

So the worst that could happen between Mastodon instances with a subject field and Mastodon instances on older versions without one is that subjects from newer versions aren't shown on older versions. During the transition phase, users on newer versions have to consider those on older versions and not rely on subjects too much. That'd be all.

**Comment by trwnh** [comment]
wait, what's the difference between a subject and a summary?

**Comment by JupiterRowland** [comment]
> wait, what's the difference between a subject and a summary?

Subject is something that e-mail has, that blogs have, that older Fediverse projects like Friendica and some newer ones have, but that Mastodon doesn't have. It's the title of a post, much like the title of a newspaper article. It always stands at the top, above everything.

Summary is what Mastodon uses for content warnings, and what Friendica uses as an actual summary. Like, a short description of a long post so that you don't get thousands of characters slammed into your face right away. Click it, and you see the post itself, and the summary disappears. Click it again, and you see the summary again, and the post text disappears. Summary is basically a "tl;dr" that you can click to what's tl;dr.

**Comment by sjehuda** [comment]
This could be related to ticket #22641 (adding titles to feeds).

**Comment by shaedrich** [comment]
People abusing the CWs as subject is one scenario, but another one I can think of is when people use CWs that are not necessarily descriptive of the content, so in the collapsed state, one is only able to tell, what possible "triggers" within the text are, but not what the text is about as a whole, because the text to make that determination is not shown.

**Comment by JupiterRowland** [comment]
> People abusing the CWs as subject is one scenario

If it's a _summary_ and not a _subject_, it isn't necessarily abuse. There are projects in the Fediverse that have always used this field for summaries, and "always" may pretty well mean "since before Mastodon".

See my comment five comments up from here.

**Comment by shaedrich** [comment]
Since the Mastodon web version calls it "Content Warning", I'd say, it's the official use in Mastodon (not necessarily the Fediverse). We could argue all day long, but I'm sure, we agree, these things should not be mixed regardless of what they are used for. Thus, there needs to be a change here.

**Comment by joyeusenoelle** [comment]
[The purpose of a system is what it does](https://en.wikipedia.org/wiki/The_purpose_of_a_system_is_what_it_does). If the "content warning" field is commonly used as a post summary/subject, it doesn't really matter what the _intent_ of the field is, except inasmuch as it demonstrates how usage has diverged from the original intent.

You say "I'm sure we agree": we don't. I remain in favor of changing the name of the field to "Subject" (which is how many people use it) or "Summary" (which is the name of the field in the AP specification) - I don't have a strong opinion on which, except that "Subject" may make new users feel that, as in email, it's required by etiquette - and making explicit, in a tooltip or hint, that it can be used for, but is not limited to, content warnings. There's no reason to add additional clutter by splitting out the subject/summary field into its own thing, especially when "Content Warning" is already a renaming of the AP summary field.

**Comment by shaedrich** [comment]
> You say "I'm sure we agree": we don't. I remain in favor of changing the name of the field to "Subject" (which is how many people use it) or "Summary" (which is the name of the field in the AP specification)

The field could be renamed. That's what I meant with "Thus, there needs to be a change". So, agreement here, indeed.

But why would someone want toots having a subject be hidden like they currently are since they are interpreted as content warnings? That doesn't make sense for subjects as you suggest.

**Comment by joyeusenoelle** [comment]
> The field could be renamed. That's what I meant with "Thus, there needs to be a change". So, agreement here, indeed.

Understood.

> But why would someone want toots having a subject be hidden like they currently are since they are interpreted as content warnings? That doesn't make sense for subjects as you suggest.

I've answered that in an earlier comment:

> In fact, [hiding posts with a toggle] remains a substantial benefit as far as I'm concerned; if someone's talking about a TV show I don't watch, I don't need a content warning for it, but I'm still not going to click the "show more" to open the full post, much like if I read the subject line of an email and decide I don't care about it, I won't click to open the email.

**Comment by shaedrich** [comment]
Yeah, but that doesn't really convince me. Some people would want that and some don't. There isn't a strong need for initially hiding subjects just in case someone doesn't like it. However, for content warnings, there's a way stronger case.

**Comment by joyeusenoelle** [comment]
Yes, the proposal I was agreeing with includes determining whether to collapse the post based on whether it's marked `sensitive`, not whether it has a subject/summary line. The two do not need to be connected.

I strongly encourage you to re-read the comment thread. Most of your questions have been answered already.

**Comment by shaedrich** [comment]
> Yes, the proposal I was agreeing with includes determining whether to collapse the post based on whether it's marked `sensitive`, not whether it has a subject/summary line.

Currently, the web client doesn't offer to mark a toot as sensitive separately, as far as I know. Of course, implementing this as a checkbox, would help.

> The two do not need to be connected.

However, I wouldn't recommend storing a string in one field and a boolean in another and to know if the string is X or Y, you need to check the boolean. This is, as I said, mixing of data is not a good way to implement something.

If you say, the text field is always a subject and the collapsing is done by `sensitive`, there technically still wouldn't be a content warning and people wouldn't know, what they are warned about. Content warnings aren't trivial. People rely on them related to their psychological conditions. Ideally, content warnings could be filtered based on settings, but that's another issue. However, filtering would be hard, if content warnings are arbitrary text and encouraged (by the label) to be fancy, "clever", lengthy, text.

**Comment by JupiterRowland** [comment]
> But why would someone want toots having a subject be hidden like they currently are since they are interpreted as content warnings? That doesn't make sense for subjects as you suggest.

A summary for a maximum of 500 characters doesn't make sense. That's why Mastodon didn't use this field at all at first.

But it could still be used for content warnings by those who got used to it, and who don't want to switch to the actually older way of having content warnings generated reader-side by word filters and having keywords or hashtags in your posts that trigger these filters.

Also, Mastodon doesn't know spoiler tags which there isn't even any actual ActivityPub standard for and which don't exist in Rich Text and standard Markdown either, not to mention that they require JavaScript. So regardless of what it's named, it can stand in for a spoiler tag.

Still, I prefer "summary" over "subject". "Summary" is the official name, and "subject" will cause continued confusion with the title field on Friendica, Hubzilla and (streams).

**Comment by JupiterRowland** [comment]
> However, for content warnings, there's a way stronger case.

Mastodon's way of content warnings is often criticised for forcing itself upon everyone: poster-side content warnings, and everyone who receives the post is being warned all the same, regardless of whether they need a warning or not. Or if someone can't be bothered to add a CW, or if there's a technological barrier that keeps someone from adding a CW (no summary field for comments on Hubzilla/(streams), no summary field at all on Friendica), then there's no CW for anyone.

Maybe, now that Mastodon has filters that can hide posts behind automatically generated CWs, Mastodon could transition to these as a default and thus pick up a method that has been in the Fediverse for longer than Mastodon itself: reader-side content warnings, automatically generated individually for each reader by word filters, depending on what they need to be warned about.

That wouldn't even mean an extra effort for the posters. They'd just have to get used to having certain keywords or hashtags in their posts so that sensitive users' filters are triggered.

The two major obstacles are that poster-side CWs in the summary field are firmly engrained in Mastodon's culture, and filters are too difficult to use and actually completely unknown to many users.

I've actually proposed easier-to-use, preset-based filters for sensitive content (#26259), but to no avail so far.

**Comment by shaedrich** [comment]
> Also, Mastodon doesn't know spoiler tags which there isn't even any actual ActivityPub standard for and which don't exist in Rich Text and standard Markdown either, not to mention that they require JavaScript. So regardless of what it's named, it can stand in for a spoiler tag.

Spoiler tags would be another way to implement this. However, [`<details>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) doesn't require JavaScript.
